segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -132,240 +160,352 @@ public GarbageCollector(int containerId, ChunkStorage chunkStorage,
+        this.taskScheduler = new MultiKeySequentialProcessor<>(storageExecutor);
     }
 
     /**
      * Initializes this instance.
+     * @param taskQueue Task queue to use.
      */
-    public void initialize() {
-        Services.startAsync(this, this.executor);
+    public CompletableFuture<Void> initialize(AbstractTaskQueueManager<TaskInfo> taskQueue) {
+        this.taskQueue = Preconditions.checkNotNull(taskQueue, "taskQueue");
+        return taskQueue.addQueue(this.taskQueueName, false)
+                .thenComposeAsync(v -> taskQueue.addQueue(this.failedQueueName, true), storageExecutor);
     }
 
     /**
-     * Gets a value indicating how much to wait for the service to shut down, before failing it.
+     * Adds given chunks to list of garbage chunks.
      *
-     * @return The Duration.
+     * @param chunksToDelete List of chunks to delete.
      */
-    @Override
-    protected Duration getShutdownTimeout() {
-        return SHUTDOWN_TIMEOUT;
+    CompletableFuture<Void> addChunksToGarbage(long transactionId, Collection<String> chunksToDelete) {
+        Preconditions.checkState(null != taskQueue, "taskQueue must not be null.");
+        val futures = new ArrayList<CompletableFuture<Void>>();
+        val startTime = currentTimeSupplier.get() + config.getGarbageCollectionDelay().toMillis();
+
+        chunksToDelete.forEach(chunkToDelete -> futures.add(addChunkToGarbage(transactionId, chunkToDelete, startTime, 0)));
+        return Futures.allOf(futures);
     }
 
     /**
-     * Main execution of the Service. When this Future completes, the service auto-shuts down.
+     * Adds given chunk to list of garbage chunks.
      *
-     * @return A CompletableFuture that, when completed, indicates the service is terminated. If the Future completed
-     * exceptionally, the Service will shut down with failure, otherwise it will terminate normally.
+     * @param chunkToDelete Name of the chunk to delete.
+     * @param startTime Start time.
+     * @param attempts Number of attempts to delete this chunk so far.
      */
-    @Override
-    protected CompletableFuture<Void> doRun() {
-        loopFuture = Futures.loop(
-                this::canRun,
-                () -> delaySupplier.get()
-                        .thenComposeAsync(v -> deleteGarbage(true, config.getGarbageCollectionMaxConcurrency()), executor)
-                        .handleAsync((v, ex) -> {
-                            if (null != ex) {
-                                log.error("{}: Error during doRun.", traceObjectId, ex);
-                            }
-                            return null;
-                        }, executor),
-                executor);
-        return loopFuture;
+    CompletableFuture<Void> addChunkToGarbage(long transactionId, String chunkToDelete, long startTime, int attempts) {
+        Preconditions.checkState(null != taskQueue, "taskQueue must not be null.");
```

Comment:
`addChunkToGarbage` and `addSegmentToGarbage` seem to share almost the same logic. It looks simple to create a common method for both, and just use as input the right metric and `TaskInfo` type depending on whether it runs for a chunk or a segment.

Reply:
There are subtle differences like extra parameters like attempts, start time etc. These methods are not exactly same.
These methods are small enough and extracting method will probably will be more complicated than necessary so this apparent duplication is okay.